prow/test/integration/cmd/fakepubsub/main.go

```diff
@@ -0,0 +1,186 @@
+	emulatorHostPort string
+}
+
+func (o *options) validate() error {
+	return nil
+}
+
+func flagOptions() *options {
+	// When the KIND cluster starts, the Prow configs get loaded into a
+	// Kubernetes ConfigMap object. This object is then mounted as a volume into
+	// the fakepubsub container at the path "/etc/config/config.yaml".
+	o := &options{config: configflagutil.ConfigOptions{ConfigPath: "/etc/config/config.yaml"}}
+	fs := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+
+	fs.StringVar(&o.emulatorHostPort, "emulator-host-port", "0.0.0.0:8085", "Host and port of the running Pub/Sub emulator.")
+	o.config.AddFlags(fs)
+
+	fs.Parse(os.Args[1:])
+
+	return o
+}
+
+func main() {
+	logrusutil.ComponentInit()
+
+	o := flagOptions()
+	if err := o.validate(); err != nil {
+		logrus.WithError(err).Fatal("Invalid arguments.")
+	}
+
+	health := pjutil.NewHealth()
+	health.ServeReady()
+
+	defer interrupts.WaitForGracefulShutdown()
+
+	if err := startPubSubEmulator(o); err != nil {
+		logrus.WithError(err).Fatal("could not start Pub/Sub emulator")
+	}
+
+	if err := initEmulatorState(o); err != nil {
+		logrus.WithError(err).Fatal("Could not initialize emulator state")
+	}
+}
+
+// startPubSubEmulator starts the Pub/Sub Emulator. It's a Java server, so the
+// host system needs the JRE installed as well as gcloud cli (this the
+// recommended way to start the emulator).
+func startPubSubEmulator(o *options) error {
+	logrus.Info("Starting Pub/Sub emulator...")
+
+	args := []string{"beta", "emulators", "pubsub", "start",
+		fmt.Sprintf("--host-port=%s", o.emulatorHostPort)}
+	cmd := exec.Command("gcloud", args...)
+
+	// Unfortunately the emulator does not really give useful messages about
+	// what type of gRPC request is being served. Still, this is better than
+	// nothing.
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+	if err := cmd.Start(); err != nil {
+		return fmt.Errorf("Could not start process: %v", err)
+	}
+	logrus.Info("Started Pub/Sub emulator")
+
+	pgid, err := syscall.Getpgid(cmd.Process.Pid)
+	if err != nil {
+		return fmt.Errorf("Could not get pid: %v", err)
+	}
+
+	// Cleanup. Kill child processes (in our case, the emulator) if we detect
+	// that we're getting shut down. See
+	// https://stackoverflow.com/a/29552044/437583.
+	interrupts.Run(func(ctx context.Context) {
+		for {
+			if _, ok := <-ctx.Done(); ok {
+				syscall.Kill(-pgid, syscall.SIGTERM)
+				cmd.Wait()
+				logrus.Info("Pub/Sub emulator exited.")
+				return
+			}
+		}
+	})
+
+	return nil
+}
+
+// initEmulatorState creates Pub/Sub topics and subscriptions, because
+// every time the emulator starts, it starts off from a clean slate (no topics
+// or subscriptions).
+func initEmulatorState(o *options) error {
+	configAgent, err := o.config.ConfigAgent()
+	if err != nil {
+		return fmt.Errorf("Error starting config agent: %v", err)
+	}
+
+	subs := configAgent.Config().PubSubSubscriptions
```

Comment:
> my initial reaction to initialize pubsub topics is that it should be done by client(tests) instead of server

I agree with your taste. But the additional complexity that is hidden here is sub's own startup code to set up its PullServer. If we don't  run the topic/subscription creation ASAP (which is what we're doing here), `sub` will crashloop because fakepubsub will tell sub that the subscription it wants to subscribe to does not exist. So if we move this to the test case, sub will not be in a "ready" state (which we wait for in setup-prow-components.sh) and we will fail the integration test setup phase before we even get to the tests.